src/control/server/mgmt_svc.go

```diff
@@ -96,3 +133,136 @@ func (svc *mgmtSvc) checkReplicaRequest(req proto.Message) error {
 	}
 	return svc.checkSystemRequest(req)
 }
+
+// startBatchLoops kicks off the asynchronous batch processing loops.
+func (svc *mgmtSvc) startBatchLoops(ctx context.Context) {
+	go svc.joinLoop(ctx)
+	go svc.batchLoop(ctx)
+}
+
+// submitBatchRequest submits a message for batch processing and waits for a response.
+func (svc *mgmtSvc) submitBatchRequest(ctx context.Context, msg proto.Message) (proto.Message, error) {
+	respCh := make(batchRespChan, 1)
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case svc.batchReqs <- &batchRequest{msg: msg, ctx: ctx, respCh: respCh}:
+	}
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case resp := <-respCh:
+		return resp.msg, resp.err
+	}
+}
+
+// processBatchPoolEvictions processes a batch of PoolEvictReq messages,
+// consolidating them by pool ID in order to minimize the number of
+// dRPC calls to the engine.
+func (svc *mgmtSvc) processBatchPoolEvictions(ctx context.Context, bprChan batchProcessRespChan, reqs []*batchRequest) []*batchRequest {
+	poolReqs := make(map[string][]*batchRequest)
+
+	for _, req := range reqs {
+		msg, ok := req.msg.(*mgmtpb.PoolEvictReq)
+		if !ok {
+			svc.log.Errorf("unexpected message type %T", req.msg)
+			continue
+		}
+		poolReqs[msg.Id] = append(poolReqs[msg.Id], req)
+	}
+
+	for poolID, reqs := range poolReqs {
+		if len(reqs) == 0 {
+			continue
+		}
+
+		batchReq := &mgmtpb.PoolEvictReq{Id: poolID, Sys: reqs[0].msg.(*mgmtpb.PoolEvictReq).Sys}
+		for _, req := range reqs {
+			batchReq.Handles = append(batchReq.Handles, req.msg.(*mgmtpb.PoolEvictReq).Handles...)
+		}
+
+		resp, err := svc.evictPoolConnections(ctx, batchReq)
+		if err != nil {
+			svc.log.Errorf("failed to evict pool connections for pool %s: %s", poolID, err)
+			for _, req := range reqs {
+				req.sendResponse(ctx, nil, err)
+			}
+			continue
+		}
+		for _, req := range reqs {
+			req.sendResponse(ctx, resp, nil)
+		}
+	}
+
+	return nil
+}
+
+// processBatchedMsgRequests processes a batch of requests for a given message type.
+func (svc *mgmtSvc) processBatchedMsgRequests(ctx context.Context, bprChan batchProcessRespChan, msgName string, reqs []*batchRequest) {
+	bpr := &batchProcessResp{msgName: msgName}
+	if len(reqs) == 0 {
+		select {
+		case <-ctx.Done():
+			return
+		case bprChan <- bpr:
+		}
+		return
+	}
+
+	switch msgName {
+	case string(proto.MessageName(new(mgmtpb.PoolEvictReq))):
+		bpr.retryableReqs = svc.processBatchPoolEvictions(ctx, bprChan, reqs)
+	default:
+		svc.log.Errorf("no batch handler for message type %s", msgName)
```

Comment:
is the reason we are allowing for this related to compatibility reasons with different software versions that might support different message types?

Reply:
More that the other option is a `panic()`, and that seems overly harsh. I suppose the argument could be made that a panic like that could only happen in development, though, so maybe it's fine.